src/global_distrib/mod_global_distrib.erl

```diff
@@ -0,0 +1,185 @@
+                    drop
+            end;
+
+        error ->
+            ?DEBUG("Unable to route global message id=~s from=~s to=~s: "
+                   "user not found in the routing table",
+                   [get_metadata(Acc, id), jid:to_binary(From), jid:to_binary(To)]),
+            ejabberd_hooks:run_fold(mod_global_distrib_unknown_recipient,
+                                    GlobalHost, {From, To, Acc, Packet}, [])
+    end.
+
+%%--------------------------------------------------------------------
+%% Helpers
+%%--------------------------------------------------------------------
+
+-spec maybe_initialize_metadata(mongoose_acc:t()) -> mongoose_acc:t().
+maybe_initialize_metadata(Acc) ->
+    case mongoose_acc:get(global_distrib, Acc, undefined) of
+        #{} -> Acc;
+        undefined ->
+            Metadata = #{ttl => opt(message_ttl),
+                         id => uuid:uuid_to_string(uuid:get_v4(), binary_standard),
+                         origin => opt(local_host)},
+            mongoose_acc:put(global_distrib, Metadata, Acc)
+    end.
+
+-spec get_bound_connection(Server :: ejabberd:lserver()) -> pid().
+get_bound_connection(Server) ->
+    get_bound_connection(Server, get({connection, Server})).
+
+-spec get_bound_connection(Server :: ejabberd:lserver(), pid() | undefined) -> pid().
+get_bound_connection(Server, undefined) ->
+    Pid = mod_global_distrib_sender:get_process_for(Server),
+    put({connection, Server}, Pid),
+    Pid;
+get_bound_connection(Server, Pid) when is_pid(Pid) ->
+    case is_process_alive(Pid) of
+        false -> get_bound_connection(Server, undefined);
+        true -> Pid
+    end.
+
+-spec deps(Opts :: proplists:proplist()) -> gen_mod:deps_list().
+deps(Opts) ->
+    ConnectionsOpts = proplists:get_value(connections, Opts, []),
+    CacheOpts = proplists:get_value(cache, Opts, []),
+    BounceOpts = proplists:get_value(bounce, Opts, []),
+
+    Deps0 = [{mod_global_distrib_mapping, CacheOpts ++ Opts, hard},
+             {mod_global_distrib_disco, Opts, hard},
+             {mod_global_distrib_receiver, ConnectionsOpts ++ Opts, hard},
+             {mod_global_distrib_sender, ConnectionsOpts ++ Opts, hard}],
+    case BounceOpts of
+        false -> Deps0;
+        _ -> [{mod_global_distrib_bounce, BounceOpts ++ Opts, hard} | Deps0]
+    end.
+
+-spec start() -> any().
+start() ->
+    mongoose_metrics:ensure_metric(global, ?GLOBAL_DISTRIB_DELIVERED_WITH_TTL, histogram),
+    mongoose_metrics:ensure_metric(global, ?GLOBAL_DISTRIB_STOP_TTL_ZERO, spiral),
+    ejabberd_hooks:add(filter_packet, global, ?MODULE, maybe_reroute, 99).
+
+-spec stop() -> any().
+stop() ->
+    ejabberd_hooks:delete(filter_packet, global, ?MODULE, maybe_reroute, 99).
+
+-spec lookup_recipients_host(jid(), binary(), binary()) -> {ok, binary()} | error.
+lookup_recipients_host(#jid{luser = <<>>, lserver = LServer}, LocalHost, GlobalHost)
```

Comment:
You skipped the guards: `when LServer == LocalHost; LServer == GlobalHost`

Reply:
Actually I didn't. Let me rephrase: Why do we always handle locally messages addressed to GlobalHost? Probably I'm just missing something.

Reply:
There are no mappings that map JID `globalhost` to anything. It isn't a user (because of the empty user part), it isn't a component (it would have its own domain). Valid contents of such packets are pretty much limited to `iq` stanzas that require a special treatment per XEP (for example, disco [which is local] returns all components registered at any cluster). Any cluster is as capable of handling such messages as others.

Reply:
Or phrasing it another way, `globalhost` is this cluster as much as any other cluster, so might as well handle it here.